Sources/Foundation/JSONSerialization.swift

```diff
@@ -549,523 +615,913 @@ private struct JSONWriter {
+                    continue
+                }
+
+                guard let extraByte = reader.value else {
+                    throw JSONError.unexpectedEndOfFile
+                }
+
+                switch extraByte {
+                case UInt8(ascii: ","):
+                    state = .expectKey
+                case UInt8(ascii: "}"):
+                    return object
+                case UInt8(ascii: " "), UInt8(ascii: "\r"), UInt8(ascii: "\n"), UInt8(ascii: "\t"):
+                    state = .expectSeperatorOrEnd
+                default:
+                    throw JSONError.unexpectedCharacter(ascii: extraByte, characterIndex: self.reader.index)
+                }
 
-            if JSONReader.oneToNine.contains(ascii) {
-                guard let ch = readDigits() else { return true }
-                ascii = ch
-                if [ JSONReader.DECIMAL_SEPARATOR, JSONReader.LOWER_EXPONENT, JSONReader.UPPER_EXPONENT ].contains(ascii) {
-                    guard nextASCII() else { return false } // There should be at least one char as readDigits didn't remove the '.eE'
+            case .expectSeperatorOrEnd:
+                seperatorloop: while let (byte, index) = reader.read() {
+                    switch byte {
+                    case UInt8(ascii: " "), UInt8(ascii: "\r"), UInt8(ascii: "\n"), UInt8(ascii: "\t"):
+                        continue
+                    case UInt8(ascii: "}"):
+                        return object
+                    case UInt8(ascii: ","):
+                        state = .expectKey
+                        break seperatorloop
+                    default:
+                        throw JSONError.unexpectedCharacter(ascii: byte, characterIndex: index)
+                    }
+                }
+
+                guard case .expectKey = state else {
+                    throw JSONError.unexpectedEndOfFile
                 }
-            } else if ascii == JSONReader.ZERO {
-                guard nextASCII() else { return true }
+            case .expectKeyOrEnd:
+                preconditionFailure("this state should be unreachable here")
+            }
+        }
+    }
+}
+
+extension JSONParserImpl {
+    
+    @usableFromInline struct DocumentReader {
+        @usableFromInline let array: [UInt8]
+        @usableFromInline let count: Int
+
+        @usableFromInline /* private(set) */ var index: Int = -1
+        @usableFromInline /* private(set) */ var value: UInt8?
+
+        @inlinable init<Bytes: Collection>(bytes: Bytes) where Bytes.Element == UInt8 {
+            if let array = bytes as? [UInt8] {
+                self.array = array
             } else {
-                throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.propertyListReadCorrupt.rawValue,
-                              userInfo: [NSDebugDescriptionErrorKey : "Numbers must start with a 1-9 at character \(input)." ])
+                self.array = Array(bytes)
             }
```

Comment:
Does `DocumentReader` need to take the input as ` <Bytes: Collection>` ? It looks like the input is always an `UnsafeBufferPointer<UInt8>` and simply storing that struct may be faster. Calling `Array(bytes)` or `as? [UIntt8]` may end up allocating another buffer. I think the generics are just going to add overhead here.

Reply:
I'm pretty sure the compiler will create two perfectly optimized methods for this. Which means we won't have any performance hits here. I don't think we should operate on `UnsafeBufferPointer<UInt8>` here, since this comes with Security implications, based on the unsafe memory access in release builds.